compiler/rustc_trait_selection/src/traits/select/candidate_assembly.rs

```diff
@@ -320,6 +321,34 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
         }
     }
 
+    /// Implements one of the `Fn()` family for a fn pointer.
+    #[instrument(skip(self, candidates), level = "debug")]
+    fn assemble_builtin_callable_candidates(
+        &mut self,
+        obligation: &TraitObligation<'tcx>,
+        candidates: &mut SelectionCandidateSet<'tcx>,
+    ) {
+        trace!(substs = ?obligation.predicate.skip_binder().trait_ref.substs);
+        // Okay to skip binder because what we are inspecting doesn't involve bound regions.
+        let self_ty = obligation.self_ty().skip_binder();
+        trace!(?self_ty);
+        trace!(ty = ?self_ty.kind());
+        match *self_ty.kind() {
+            ty::Infer(ty::TyVar(_)) => {
+                debug!("assemble_builtin_callable_candidates: ambiguous self-type");
+                candidates.ambiguous = true; // Could wind up being a fn() type.
+            }
+            // Unsafe or extern functions do not necessarily implement FnOnce,
+            // but they do implement `Callable`
+            ty::FnDef(..) | ty::FnPtr(..) | ty::Closure(..) => {
+                candidates.vec.push(BuiltinCandidate { has_nested: false })
```

Comment:
This incorrectly implements `fn(): Callable<(i32,)>` doesn't it?

Reply:
```rust
#![feature(fn_traits)]

fn test<T: std::ops::Callable<()>>() {

}

fn main() {
    test::<fn(i32) -> i32>();
}
```

That passes on the current solver. See `confirm_fn_pointer_candidate` for how to extract the args tuple + `SelectionContext::confirm_poly_trait_refs` to confirm.